dataframe.py

```diff
@@ -0,0 +1,269 @@
+from abc import ABC, abstractmethod
+from collections.abc import Mapping
+from typing import Any, Iterable, Sequence
+
+# ----------------------------------------------------------------------
+# A simple data type class hierarchy for illustration
+
+
+class DataType(ABC):
+    """
+    A metadata object representing the logical value type of a cell in a data
+    frame column. This metadata does not guarantee an specific underlying data
+    representation
+    """
+    def __eq__(self, other: 'DataType'):
+        return self.equals(other)
+
+    def __str__(self):
+        return self.to_string()
+
+    def __repr__(self):
+        return str(self)
+
+    @abstractmethod
+    def to_string(self) -> str:
+        """
+        Return human-readable representation of the data type
+        """
+
+    @abstractmethod
+    def equals(self, other: 'DataType') -> bool:
+        """
+        Return true if other DataType contains the same metadata as this
+        DataType
+        """
+        pass
+
+
+class PrimitiveType(DataType):
+
+    def equals(self, other: DataType) -> bool:
+        return type(self) == type(other)
+
+
+class NullType(PrimitiveType):
+    """
+    A data type whose values are always null
+    """
+    def to_string(self):
+        return "null"
+
+
+class Boolean(PrimitiveType):
+
+    def to_string(self):
+        return "bool"
+
+
+class NumberType(PrimitiveType):
+    pass
+
+
+class IntegerType(NumberType):
+    pass
+
+
+class SignedIntegerType(IntegerType):
+    pass
+
+
+class Int8(SignedIntegerType):
+
+    def to_string(self):
+        return "int8"
+
+
+class Int16(SignedIntegerType):
+
+    def to_string(self):
+        return "int16"
+
+
+class Int32(SignedIntegerType):
+
+    def to_string(self):
+        return "int32"
+
+
+class Int64(SignedIntegerType):
+
+    def to_string(self):
+        return "int64"
+
+
+class Binary(PrimitiveType):
+    """
+    A variable-size binary (bytes) value
+    """
+    def to_string(self):
+        return "binary"
+
+
+class String(PrimitiveType):
+    """
+    A UTF8-encoded string value
+    """
+    def to_string(self):
+        return "string"
+
+
+class Object(PrimitiveType):
+    """
+    Any PyObject value
+    """
+    def to_string(self):
+        return "object"
+
+
+class Categorical(DataType):
+    """
+    A categorical value is an ordinal (integer) value that references a
+    sequence of category values of an arbitrary data type
+    """
+
+    def __init__(self, index_type: IntegerType, category_type: DataType,
+                 ordered: bool = False):
```

Comment:
I'd like to understand better the reason to have the types of the data and the categories in the type.

With an example, I think a category column could be represented with something equivalent to:
```python
class CategoryColumn:
    dtype = 'category'
    index = pyarrow.array([0, 1, 0, 2, 0, 1], type='int8')
    categories = pyarrow.array(['red', 'green', 'blue'], type='string')
```
So, when accessing the data, I can imagine that we may be interested in one of two options:
1. The underlying data (to copy the structure from implementation to implementation for example)
2. To access the actual data (forgetting about the internal representation)

For the first option, an example of code copying the data to a numpy array could be:
```python
for source_column in dataframe:
    if source_column.dtype == 'category':
        copy_of_index = numpy.array(source_column.index, dtype=source_column.dtype)
```
The example is very simplistic, but it shows that we can get the type of the index (`int8` in the previous example) from the underlying column (and not from the data type). It feels like if we have it again in the data type it will just be a duplicate that will be need to keep in sync. Same would apply to the data type of the categories.

What would be the main advantage of this implementation, as opposed to have just a categorical type with only the `ordered` parameter (or two types `categorical` and `ordered_categorical`)? I guess it decouples the schema from the data, but I'm unsure if it's worth the extra complexity. I guess we could just identify data types by their name as a string among implementations, instead of making them implement all these subclasses, if we don't have parameters. Which I guess would make things simpler.

Am I missing something? Does it make sense what I say?

Reply:
It's pretty common for users to compare dtypes across columns. With this proposed change it would have a category column with `int32` codes and say `int64` categories have the same dtype as `int8` codes and `string` categories, which may lead to confusion / frustration for end users.

I'm a +1 to including the index / categories types in the dtype as it makes things more explicit for end users.

Reply:
Thanks, that's useful to know. Just to understand better, do you have any example on why users compare dtypes across columns?

Reply:
There's lots of reasons to have access to the complete schema, from preallocating space for data (in-memory or on disk) to building query plans (including kernel selection) before data arrives. Not having access to the index type and category type would (IMHO) make a lot of things much more difficult. 

That said, if you have an application where the index or category type is simply not known, you could use an `Any` type to indicate that the actual type must be observed in the actual data

Reply:
> Thanks, that's useful to know. Just to understand better, do you have any example on why users compare dtypes across columns?

Unfortunately our users don't share the "why" too often, but one use case is checking the output of dtype inference from something like a `read_csv` call.

Reply:
> Not having access to the index type and category type would (IMHO) make a lot of things much more difficult.

Thanks @wesm, that makes sense. But I don't fully understand this part. Given a categorical column, there is access to the index and category types in my example (there was a typo in my previous code, may be that's where I didn't make myself clear). It's just the place where it is found that is different (in the array instead of in the dtype).

In your proposal:
```python
my_column.dtype.index_type
```
In my example would be something like:
```python
my_column.underlying_categorical_data.index.dtype
```

I understand that in my example there wouldn't be access to the underlying types if a dataframe schema is used without the actual dataframe. Then, with your proposal we could still know the internal types of the categorical, while in my example those will be undefined, since they should be in the data, which doesn't exist. Is this what you have in mind? If it is, in what case are we considering that a schema can live without the data?

Reply:
Accessing the actual data is potentially costly, I think? So that would make getting the type of the categories costly, while I assume accessing it from the dtype itself would not.

Reply:
> Accessing the actual data is potentially costly, I think?

Good point. In the PR, the `type` (aka `dtype`) is a property of the `Column` class, so accessing the index type would be
```python
class Column:
    @property
    def type(self):
        pass

my_column.type.index_type
```

The data is not specified, may be that's the problem, but if we agree the categorical data as two subcolumns, one for the index, one for the categories, then accessing the index type with an unparametrized categorical dtype would be:
```
my_column.index_subcolumn.type
```
I may be missing something, but I would say in both cases the type is metadata of the `Column` class that doesn't require accessing the actual data. Not sure if in certain architectures (distributed, GPU...) accessing a subcolumn could be expensive, but in pure Python both cases would be two lookups in the namespace.

I'm not sure what would imply if we consider implementations using Arrow's `Dict` instead of the two arrays separately. Would that be costly?